packages/slate/src/interfaces/range.ts

```diff
@@ -0,0 +1,269 @@
+import { produce } from 'immer'
+import isPlainObject from 'is-plain-object'
+import { Operation, Path, Point, PointEntry } from '..'
+
+/**
+ * `Range` objects are a set of points that refer to a specific span of a Slate
+ * document. They can define a span inside a single node or a can span across
+ * multiple nodes.
+ */
+
+export interface Range {
+  anchor: Point
+  focus: Point
+  [key: string]: any
+}
+
+export const Range = {
+  /**
+   * Get the start and end points of a range, in the order in which they appear
+   * in the document.
+   */
+
+  edges(
+    range: Range,
+    options: {
+      reverse?: boolean
+    } = {}
+  ): [Point, Point] {
+    const { reverse = false } = options
+    const { anchor, focus } = range
+    return Range.isBackward(range) === reverse
+      ? [anchor, focus]
+      : [focus, anchor]
+  },
+
+  /**
+   * Get the end point of a range.
+   */
+
+  end(range: Range): Point {
+    const [, end] = Range.edges(range)
+    return end
+  },
+
+  /**
+   * Check if a range is exactly equal to another.
+   */
+
+  equals(range: Range, another: Range): boolean {
+    return (
+      Point.equals(range.anchor, another.anchor) &&
+      Point.equals(range.focus, another.focus)
+    )
+  },
+
+  /**
+   * Check if a range exists in a list or map of ranges.
+   */
+
+  exists(range: Range, target: Range[] | Record<string, Range>): boolean {
+    if (Range.isRangeList(target)) {
+      return !!target.find(r => Range.equals(r, range))
+    }
+
+    if (Range.isRangeMap(target)) {
+      for (const key in target) {
+        if (Range.equals(range, target[key])) {
+          return true
+        }
+      }
+    }
+
+    return false
+  },
+
+  /**
+   * Check if a range includes a path, a point or part of another range.
+   */
+
+  includes(range: Range, target: Path | Point | Range): boolean {
+    if (Range.isRange(target)) {
+      if (
+        Range.includes(range, target.anchor) ||
```

Comment:
Shouldn't this be '&&'? Or the next isBefore and isAfter should reverse parameters' order?

Reply:
I don't think so. First it checks whether either anchor or focus is inside `range` to cover overlap cases. If that isn't the case, it checks whether `target` is wholly inside `range`.

Reply:
then, it does not cover the range is wholly inside the target case. which shouldn't be also the "part of another range" case?

Reply:
Hmm you're right it probably needs both `range includes target` and `target includes range` checks. There aren't any tests for this either, only [the path and point scenarios](https://github.com/ianstormtaylor/slate/blob/master/packages/slate/test/interfaces/Range/includes). Would you mind logging a new ticket about it?